1.0.0 refactor

antonivanopoulos opened the pull request:
PR involves a major rethink about how we're handling deliveries and aggregation of notifications.

Currently, the base `Notification` class knows way too much about how aggregation works, as well as when to schedule in the delivery of itself (adding onto that the different `message` methods that use view helpers to render the actual notification).

What's been done here is break out those different concepts into their own classes. We have the following:
- A presenter class that's used purely for rendering the notification text. 
- An aggregator class; that, given a particular notification and a set of delay intervals, will calculate which delay to use given the number of existing unread notifications in the same group.
- A scheduler class that will use an aggregator, and decide whether or not a given notification needs to be delivered at all. If so, it will create a `Delivery` record for each channel the notification is subscribed to.

The new `Delivery` model represents a delivery to be made of a given notification to a given channel. If a notification class is subscribed to both APNS and GCM, a new notification of that type will end up with two deliveries, one for APNS and one for GCM. When the delivery happens for each channel, the send time can be recorded on the delivery itself, which saves us from the issue currently where a notification gets marked as sent, despite one of those channels potentially failing.

antonivanopoulos committed:
Pulls out GCM delivery into its very own channel

Means that we have to specify two separate channels per notification to get full mobile delivery coverage. But it means that the delivery being split over two different sets of workers won't cause one platforms ability to deliver to be affected by the delivery patterns of another.

knightstick committed:
Allow Notifications to be sent to multiple channels

knightstick committed:
Only attempt to send to channels if we have any

knightstick committed:
Mark aggregate notifications as sent after notifying all channels

knightstick committed:
Add new notify_aggregated_channels! method that sets notifications to sent after delivering to all channels

knightstick committed:
Move some class methods to be private

knightstick committed:
Change class_eval to allow to receive for TestNotification

knightstick committed:
Change Channels deliver methods to use ids not notification objects

knightstick committed:
Pass the id not the notification to all channels when delivering

knightstick committed:
Mark delay_for spec as pending until it is implemented

knightstick committed:
Delay all calls to Channel#deliver

knightstick committed:
Use where instead of map(&:find) for gathering notifications from ids

knightstick committed:
Use block syntax instead of open-ended Sidekiq::Testing.fake! so it only affects one context block

knightstick committed:
Delay_for testing WIP

knightstick committed:
Implement the delay_until logic required for delaying sending of children notifications

knightstick committed:
Fall back to integer class.aggregate_per if not specifying an aggregate_per array

knightstick committed:
Cast env var connection pool size and timeout to integers

knightstick committed:
Ensure that the connection env var exists before calling to_i

knightstick committed:
Add optional 'sendable_attributes' to notification class to filter which params get send over the wire

knightstick committed:
Merge pull request #33 from Papercloud/feature/sendable-params

Add optional 'sendable_attributes' to notification class to filter wh…

knightstick committed:
Ensure we are passing the channel options when delivering aggregated notifications

The previous refactor was calling deliver and deliver_aggregated
 on a channel class without passing through the channel options,
this in particular meant that the notification sound wasn't being
passed to APNS, so notifications were arriving with the default
iOS sound, instead of the one we wanted.

By ensuring we find the particular channel's options in	`multi_deliver`
we pass through everything we know about the channel that might be
relevant

knightstick committed:
Merge pull request #34 from Papercloud/fix/channel-options

Ensure we are passing the channel options when delivering aggregated …

antonivanopoulos committed:
Updates the testing gemfiles to what is currently being used

antonivanopoulos committed:
Includes shoulda matchers for specs

antonivanopoulos committed:
Adds a presenter class to take care of formatting notifications

antonivanopoulos committed:
Adds a delivery model and worker that handles delivery

This is one of the bookends for the rewrite. The major aim for it being that we don't want notifications to deal with delivery directly at all, because it can't handle tracking the delivery steps per channel, so this model basically just represents a delivery request for a notification, to a given channel. The outcome being here that a notification setup to send to APNS and GCM for example, will end up with 2 delivery objects, and that will track the sent time.

antonivanopoulos committed:
Adds an aggregator class that deals with figuring out send times

The big change here comes to the existing notifications table. The main aim with the refactor is to not have notifications deal with delivery directly, so we don't need a state or sent timestamp. We can get restructure this to _just_ have a read timestamp, and that will dictate the state the noti needs to know about.

antonivanopoulos committed:
Adds the install step of adding the delivery migration

antonivanopoulos committed:
Adds the scheduler class for actually creating delivery objects

The scheduler takes a noti, and using the aggregator, will figure out whether it needs new delivery objects, and if so, when it needs to send those.

antonivanopoulos committed:
Cleans out the base notification class

Basically removes everything that now lives in the scheduler, aggregator and presenter classes.

antonivanopoulos committed:
Updates the notifications controller to deal with state correctly

antonivanopoulos committed:
Adds generator for test notification for dummy app

antonivanopoulos committed:
Changes the travis script as per appraisal docs

antonivanopoulos committed:
Fixing broken specs in unsubscribe

antonivanopoulos committed:
Reverses script change and ditches 1.9.3

1.9.3 has had its support dropped as of 2015 anyway.

antonivanopoulos committed:
Uses 4.2 by default

antonivanopoulos committed:
Uses a single dummy app for specs

antonivanopoulos committed:
Adds the notification js file to precompiled assets

knightstick commented on app/models/notify_user/aggregator.rb:
```diff
@@ -0,0 +1,64 @@
+    end
+
+    def delay_time_in_seconds
+      delay_time_in_minutes = aggregation_intervals[aggregation_interval] || aggregation_intervals.last
```
Could perhaps rename the `aggregration_interval` method something like `next_interval_index` - the fact that it's singular of some other thing is pretty confusing

antonivanopoulos replied:
Ah yeah, remember our conversation about this last week, will change 👍

knightstick commented on app/models/notify_user/aggregator.rb:
```diff
@@ -0,0 +1,64 @@
+      @unread_unsent_notifications_for_group ||= unread_unsent_notifications.where('notify_user_notifications.group_id = ?', notification.group_id)
+    end
+
+    # Finding the appropriate delay time:
```
You're not finding the _time_ here, just the last notification

antonivanopoulos replied:
True 👍

knightstick commented on app/models/notify_user/aggregator.rb:
```diff
@@ -0,0 +1,64 @@
+        .first
+    end
+
+    def unread_notifications_since_last_read
```
This method name is dope, makes the reading of the method body way easier

antonivanopoulos replied:
👌

knightstick commented on app/models/notify_user/base_notification.rb:
```diff
@@ -1,186 +1,27 @@
-require 'aasm'
```
👌 👌 👌

antonivanopoulos replied:
Reminds me, can take it out of the gemspec as well.

knightstick commented:
https://github.com/Papercloud/notify_user/blob/164b0b7cb87c68fe9ab7680c5564de24f55c71ec/app/models/notify_user/base_notification.rb#L150
Rubocop would tell you to use an early return guard clause here

knightstick commented on app/models/notify_user/base_notification.rb:
```diff
@@ -1,186 +1,27 @@
-    after_commit :deliver!, on: :create
-    after_commit :deliver, on: :create
-
     if ActiveRecord::VERSION::MAJOR < 4
```
This is probably redundant now, right?

antonivanopoulos replied:
Yeah 👍

knightstick commented on app/models/notify_user/base_notification.rb:
```diff
@@ -191,10 +32,9 @@ def delay_time(options)
+    self.sendable_attributes = []
 
     ## True will implement a grouping/aggregation algorithm so that even though 10 notifications are delivered eg. Push Notifications
     ## Only 1 notification will be displayed to the user within the notification.json payload
```
I think this comment could use a bit of rewording

antonivanopoulos replied:
True, but I think the whole feature needs a rethink anyway, in terms of the param grouping and what not, and how to avoid the race conditions we've run into elsewhere.

knightstick commented on app/models/notify_user/base_notification.rb:
```diff
@@ -211,190 +51,97 @@ def delay_time(options)
-              end
-            end
-          end
+        group_parents = parents_in_group.where('created_at >= ?', 24.hours.ago).order('created_at DESC')
```
Is this "reset each day" logic likely to change?

knightstick replied:
I mean, per notification type

antonivanopoulos replied:
Went with what was already there for now, but I was thinking about this as well recently. Could be something nice to add in with little effort. Could even be per channel as opposed to per notification, but either works.

knightstick commented on app/models/notify_user/base_notification.rb:
```diff
@@ -211,190 +51,97 @@ def delay_time(options)
+        group_parents = parents_in_group.where('created_at >= ?', 24.hours.ago).order('created_at DESC')
+        if group_parents.any?
+          self.parent_id = group_parents.first.id
         end
```
I'd also consider moving the "assign a parent" logic into a private method

antonivanopoulos replied:
👍

knightstick commented on app/models/notify_user/base_notification.rb:
```diff
@@ -211,190 +51,97 @@ def delay_time(options)
-        self.mark_as_sent!
-        self.class.deliver_channels(self.id)
+      save.tap do |success|
+        Scheduler.schedule(self) if deliver
```
Should this also depend on the success?

antonivanopoulos replied:
Haha yes, yes it should.

knightstick commented on app/models/notify_user/base_notification.rb:
```diff
@@ -211,190 +51,97 @@ def delay_time(options)
-        notifications = self.pending_aggregation_with(notification)
-      end
+    def parents_in_group
+      return self.class.none unless self.aggregate_grouping
```
Is this preferable to just raising an exception here? I'd say that I'd wanna know sooner rather than later than I'm using a method that makes no sense for the particular class

knightstick commented on app/models/notify_user/base_notification.rb:
```diff
@@ -211,190 +51,97 @@ def delay_time(options)
 
-      notifications.map(&:mark_as_sent)
-      notifications.map(&:save)
+    # Scopes:
```
Now might be a nice to time to move to the `class << self` syntax too, does that work for older rubies?

antonivanopoulos replied:
Should do, or at the very least, the ones being supported.

knightstick commented on app/models/notify_user/base_notification.rb:
```diff
@@ -211,190 +51,97 @@ def delay_time(options)
-        self.deliver_notifications_channel(notifications, channel_name)
-      end
+    # Configure a channel
+    def self.channel(name, options={})
```
We spoke about maybe renaming this method to `add_channel` or `register_channel`, but, I guess that's a breaking change for no real reason

antonivanopoulos replied:
Yeah, pretty much why I left it as is.

knightstick commented on app/models/notify_user/channel_presenter.rb:
```diff
@@ -0,0 +1,42 @@
+      )
+    end
+  end
+end
```
This class is awesome, I love it

knightstick commented on app/models/notify_user/delivery.rb:
```diff
@@ -0,0 +1,19 @@
+    validates :channel, presence: true
+    validates :deliver_in, presence: true
+
+    belongs_to :notification, class_name: BaseNotification.name
```
I assume this works just fine with subclasses of BaseNotification?

knightstick commented on app/models/notify_user/gcm.rb:
```diff
@@ -26,24 +26,32 @@ def valid?(payload)
+
+    def count_for_target(target)
+      BaseNotification.unread_count_for_target(target)
+    end
```
Is this duplication an argument for a `BaseService` class?

antonivanopoulos replied:
Ah, yeah, there actually is one. Moving this there was on my todos 👍

knightstick commented on app/models/notify_user/scheduler.rb:
```diff
@@ -0,0 +1,46 @@
+        aggregator = Aggregator.new(notification, options[:aggregate_per])
+        return if aggregator.has_pending_deliveries?
+
+        # Create the delivery:
```
If it needs a comment, is it better wrapped in a private method?

antonivanopoulos replied:
I don't even really think the comment is necessary. Pretty clear what the line does.

Pretty sure I just left it there from my pseudo code so everything lined up nicely.

knightstick commented on app/models/notify_user/scheduler.rb:
```diff
@@ -0,0 +1,46 @@
+    def schedule
+      notification.class.channels.each do |channel, options|
+        aggregator = Aggregator.new(notification, options[:aggregate_per])
+        return if aggregator.has_pending_deliveries?
```
I love how this gets rid of so much complexity around aggregate groupings etc

antonivanopoulos replied:
🚀

knightstick commented on app/models/notify_user/scheduler.rb:
```diff
@@ -0,0 +1,46 @@
+      last_send_time = aggregator.last_send_time
+      delay_time = aggregator.delay_time_in_seconds
+
+      validate_delay_time(last_send_time, delay_time)
```
To me, `validate_` sounds like it has side-effects rather than a return value. Maybe `validated_delay_time` is a better name?

antonivanopoulos replied:
True 👍

knightstick commented on app/serializers/notify_user/notification_serializer.rb:
```diff
@@ -5,7 +5,7 @@ class NotifyUser::NotificationSerializer < ActiveModel::Serializer
 
   def message
-    object.message
+    NotifyUser::ChannelPresenter.present(object)
```
👌 👌 👌

knightstick commented on app/workers/notify_user/delivery_worker.rb:
```diff
@@ -0,0 +1,19 @@
+      delivery = Delivery.find(delivery_id)
+
+      if delivery.notification.read?
+        # TODO: Log things out
```
Just reminding you that this is a TODO ;)

antonivanopoulos replied:
Haha 🌵

knightstick commented on app/workers/notify_user/delivery_worker.rb:
```diff
@@ -0,0 +1,19 @@
+      else
+        channel_name = delivery.channel
+        channel_options = delivery.notification.class.channels[channel_name.to_sym] || {}
+        channel_class = (channel_name + "_channel").camelize.constantize
```
This is a bit verbose, but, not really sure where it'd be best "hidden"

knightstick commented on lib/generators/notify_user/install/templates/create_notify_user_deliveries.rb:
```diff
@@ -0,0 +1,15 @@
+      t.timestamps
+    end
+
+    add_index :notify_user_deliveries, :notification_id
```
Should this have an index on `:sent_at` too, as you're querying for unsent ones?

antonivanopoulos replied:
👍

knightstick commented on lib/generators/notify_user/install/templates/create_notify_user_notifications.rb:
```diff
@@ -2,12 +2,11 @@ class CreateNotifyUserNotifications < ActiveRecord::Migration
+      t.integer :target_id
+      t.string :target_type
       t.integer :group_id
       t.integer :parent_id
```
How do you go about moving from a previous version of NU to this one? Will we need a different migration?

knightstick replied:
You also need an index on `:read_at` here I think

antonivanopoulos replied:
Yeah I think we just provide a migration snippet to run on your existing NU install to get it up to this version. That + the migration to add deliveries.

antonivanopoulos replied:
`Migrating from 0.3.2 to 1.0.0` page would work on the wiki or something.

knightstick commented on lib/notify_user/channels/action_mailer/action_mailer_channel.rb:
```diff
@@ -12,14 +12,24 @@ def default_options
+      notifications = NotifyUser::BaseNotification.where(id: notification_ids)
+
       NotifyUser::NotificationMailer.aggregate_notifications_email(notifications, default_options.deep_merge(options)).deliver
     end
```
Could this just be called from the first method? As in, if `notification_id` is an array, blah blah

antonivanopoulos replied:
Haha think this is actually your change from the GCM branch. So sure 👍 

But again...this falls into "how do we want to send aggregated notifications" territory. If params merging is the go going forward, we don't actually need this method.

knightstick commented on lib/notify_user/channels/apns/apns_channel.rb:
```diff
@@ -7,18 +7,28 @@ def default_options
 
-      NotifyUser::Apns.new(notifications, devices[:ios], options).push if devices[:ios].any?
-      NotifyUser::Gcm.new(notifications, devices[:android], options).push if devices[:android].any?
+      NotifyUser::Apns.new(notifications, devices, options).push if devices.any?
```
Do we want to log something in the case that the user doesn't have any devices?

knightstick commented on lib/notify_user/channels/gcm/gcm_channel.rb:
```diff
@@ -0,0 +1,47 @@
+  	  {
+  	    description: "Push Notifications"
+  	  }
+  	end
```
Indentation

knightstick commented on spec/lib/notify_user/channels/apns_channel_spec.rb:
```diff
@@ -11,54 +13,94 @@
+      described_class.deliver(notification.id)
+    end
+
+    it 'doesnt make use of apns' do
```
This should say `doesn\'t make use of GCM`

knightstick commented on spec/models/notify_user/aggregator_spec.rb:
```diff
@@ -0,0 +1,177 @@
+          NewPostNotification.class_eval do
+            channel :apns, aggregate_per: [0, 3, 10, 30, 60]
+            self.aggregate_grouping = true
+          end
```
Did you decide against stubbing `:channels` on the class?

antonivanopoulos replied:
Nope, just one slipped through the cracks 👍

knightstick commented on spec/models/notify_user/aggregator_spec.rb:
```diff
@@ -0,0 +1,177 @@
+          expect(aggregator.has_pending_deliveries?).to eq false
+        end
+
+        it 'returns false if there is no current pending notiifcation' do
```
typo

knightstick replied:
Or maybe it's like Desiigner

antonivanopoulos replied:
🎼 Notii notii notii notify user

knightstick commented on spec/models/notify_user/aggregator_spec.rb:
```diff
@@ -0,0 +1,177 @@
+          expect(aggregator.has_pending_deliveries?).to eq false
+        end
+
+        it 'returns false if there is pending notifications have been read' do
```
This doesn't really make sense

knightstick commented on spec/models/notify_user/aggregator_spec.rb:
```diff
@@ -0,0 +1,177 @@
+          expect(aggregator.has_pending_deliveries?).to eq false
+        end
+
+        it 'returns false if there is pending notifications have been read' do
```
Again, I think this needs the word "that"

knightstick commented on spec/models/notify_user/aggregator_spec.rb:
```diff
@@ -0,0 +1,177 @@
+      context 'without grouping' do
+        before :each do
+          NewPostNotification.class_eval do
+            channel :apns, aggregate_per: [0, 3, 10, 30, 60]
```
I'd be tempted to test with a non-zero first interval, so you know it's not short-circuiting

knightstick commented:
I didn't really take a proper look at the specs, but, overall, LGTM 🎉

antonivanopoulos committed:
Renames `aggregation_interval` to be clearer

antonivanopoulos committed:
Drops `aasm` from the dependencies

We don't need a state machine going forward

antonivanopoulos committed:
Removes the Rails < 4 compatibility stuff

antonivanopoulos committed:
Scheduling relies on the notification saving

antonivanopoulos committed:
Moves the message and count duplciates into base class

antonivanopoulos committed:
Clears up the method name for delivery time validation

antonivanopoulos committed:
Adds delivery `sent_at` index

antonivanopoulos committed:
Adds notification timestamp indexes

antonivanopoulos committed:
Removes class eval blocks from aggregator specs

antonivanopoulos committed:
Fixes apns channel typos

antonivanopoulos committed:
Cleans up aggregator spec descriptions

antonivanopoulos committed:
Fixes the copy paste error for the new indexes in notis table

antonivanopoulos committed:
Cleans up the comments for the aggregator

antonivanopoulos committed:
Moves class methods into class << self block

antonivanopoulos committed:
Fixes indentation

antonivanopoulos committed:
Fixes typo in spec name

antonivanopoulos committed:
Sets the `sent_at` timestamp during the delivery

Bit of a naive solution, we don't have a nice way to track successful / errorneous sends just yet.

antonivanopoulos committed:
Missed a `.self` when cleaning up

antonivanopoulos commented:
🚀
